theories/common.elpi

```diff
@@ -88,3 +88,24 @@ reduction-Z I O :- coq.reduction.vm.norm I {{ Z }} O, ground-Z O.
 pred negb i:bool, o:bool.
 negb tt ff :- !.
 negb ff tt :- !.
+
+pred zmod->nmod i:term, o:term.
+zmod->nmod U {{ GRing_Zmodule__to__GRing_Nmodule lp:U }} :- !.
+
+pred sring->nmod i:term, o:term.
+sring->nmod R {{ GRing_SemiRing__to__GRing_Nmodule lp:R }} :- !.
+
+pred ring->nmod i:term, o:term.
+ring->nmod R {{ GRing_Ring__to__GRing_Nmodule lp:R }} :- !.
+
+pred ring->zmod i:term, o:term.
+ring->zmod R {{ GRing_Ring__to__GRing_Zmodule lp:R }} :- !.
+
+pred ring->sring i:term, o:term.
+ring->sring R {{ GRing_Ring__to__GRing_SemiRing lp:R }} :- !.
+
+pred unitRing->ring i:term, o:term.
+unitRing->ring UR {{ GRing_UnitRing__to__GRing_Ring lp:UR }} :- !.
+
+pred field->unitRing i:term, o:term.
+field->unitRing F {{ GRing_Field__to__GRing_UnitRing lp:F }} :- !.
```

Comment:
All these names `GRing_...to...` are generated and subject to change in the future.
Can you make your code independent from these specific names?

Reply:
To be clear, these names refer to the notations you mentioned below, but the issue indeed exists (and was discussed somewhere above).

One easy way to fix it is declaring constants like:
```coq
Definition GRing_Zmodule__to__GRing_Nmodule (U : zmodType) : nmodType := U.
```
However, they are not exactly the same constants as the auto-generated subtyping functions in `ssralg.v`, and may have some consequences in the performance of conversion checking.

So, the right solution that is robust in terms of performance would be querying these constant names from the database of HB, and generating and accumulating these rules when we load Algebra Tactics. I did not bother myself checking how to do this.

Reply:
Also, Algebra Tactics is not robust regarding insertions of new structures to the hierarchy anyway. For example, MathComp 2.0 introduces the N-module and semiring structures, operators like `GRing.one`, `GRing.mul`, `GRing.natmul` are generalized to new structures, and reification rules for these operators has to be adapted to the new hierarchy. IMO, we should eventually generate reification procedure to fix these issues.

Reply:
Note that these are coercions, you should be able to do something like
```
coq.coercion.db-for (grefclass {{:gref UnitRing.type }})  (grefclass {{:gref Ring.type }}) [pr F NHoles|_]
```

Reply:
maybe even `indt GR = {{ ringType }}, coq.coercion.db-for (grefclass GR) ...`